src/extra/utils/ftd.pyx

```diff
@@ -0,0 +1,517 @@
+
+    # Interpolation is always done on double precision, see above in the
+    # implementation for CFD.
+    cdef double cmp_value = <double>ratio_value, \
+        middle_value, middle_pos = 0.0, \
+        left_pos = <double>ratio_idx, right_pos = <double>(ratio_idx + 1)
+
+    for _ in range(1, sinc_search_iterations+1):
+        middle_pos = (left_pos + right_pos) / 2
+
+        middle_value = 0.0
+        for k in range(sampling_start, sampling_end):
+            middle_value += <double>signal[k] * _sinc(k - middle_pos)
+
+        if (
+            (not negative and middle_value < cmp_value) or
+            (negative and middle_value > cmp_value)
+        ):
+            left_pos = middle_pos
+        else:
+            right_pos = middle_pos
+
+    return <data_t>middle_pos
+
+
+def cfd(
+    data_array_t signal, data_t threshold, data_t delay,
+    int width=0, data_t fraction=1.0, data_t walk=0.0,
+    int interp=EdgeInterpolation.LINEAR,
+    data_array_t edges = None,
+    data_array_t amplitudes = None
+):
+    """Constant fraction discriminator.
+
+    The pulse shape is assumed to always grow away from zero, i.e. the
+    rising slope of positive pulses is positive and that of negative
+    pulses is negative. Correspondigly, a positive threshold implies the
+    pulse to peak at its largest value while a negative threshold
+    implies the pulse to peak at its smallest value.
+
+    Args:
+        signal (array_like): 1D input array with analog signal.
+        threshold (data-type): Trigger threshold, positive values imply
+            a positive pulse slope while negative values correspondingly
+            imply a negative pulse slope.
+        delay (int): Delay between the raw and inverted signal.
+        width (int, optional): Minimal distance between found edges,
+            none by default.
+        fraction (data-type, optional): Fraction of the inverted signal,
+            1.0 by default.
+        walk (data-type, optional): Point of intersection in the
+            inverted signal, 0.0 by default.
+        interp (EdgeInterpolation, optional): Interpolation mode to
+            locate the edge position, linear by default.
+        edges (ArrayLike, optional): 1D output array to hold the
+            positions of found edges, a new one is allocated if None is
+            passed.
+        amplitudes (ArrayLike, optional): 1D output array to hold the
+            pulse amplitudes corresponding to found edges, a new one is
+            allocated if None is passed.
+
+    Returns:
+        (ArrayLike, ArrayLike, int) 1D arrays containing the edge
+            positions and amplitudes, number of found edges.
+    """
+
+    if edges is None:
+        edges = np.zeros(
+            len(amplitudes) if amplitudes is not None else len(signal) // 100,
+            dtype=np.asarray(signal).dtype)
+
+    if amplitudes is None:
+        amplitudes = np.zeros_like(edges, dtype=np.asarray(signal).dtype)
```

Comment:
I think the `amplitudes` array is never actually filled.

Reply:
Oh joy, I forgot to carry over that part of the implementation. Thanks!